### PR TITLE
Skip set gc life time

### LINKED ISSUE #3386: br: There is no need to set tikv_gc_life_time during backup and recovery
br support automatic setting `tikv_gc_life_time` during backup and recovery now. Recently there is a related [bug fix](https://github.com/pingcap/br/pull/553). It's not included in v4.0.7.

### PR DESCRIPTION
**Pull request by namco1992**
### What problem does this PR solve? 

Fix #3386.

### What is changed and how does it work?

1. Change CR structure in types.go as below:
"From TiDBAccessConfig `json:"from,omitempty"`" to "From *TiDBAccessConfig `json:"from,omitempty"`"
"To TiDBAccessConfig `json:"to,omitempty"`" to "To *TiDBAccessConfig `json:"to,omitempty"`"
2. Check `spec.from` and `spec.to` during the backup/restore, if not nil, go with the current logic to set GC; if nil, skip the GC setting logic

### Check List 

Tests 

 - Unit test
 - E2E test
 - Stability test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has Go code change

Side effects

N/A

Related changes

 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
NONE
```

### CODE REVIEW

**Comment by namco1992** [comment]
@DanielZhangQD Unfortunately I faced the same issue as @sstubbs in https://github.com/pingcap/tidb-operator/pull/3134. I'm not able to regenerate the deepcopy and tried whatever @sstubbs tried too. Does any of you know how can I solve this issue?

**Comment by DanielZhangQD** [comment]
> @DanielZhangQD Unfortunately I faced the same issue as @sstubbs in #3134. I'm not able to regenerate the deepcopy and tried whatever @sstubbs tried too. Does any of you know how can I solve this issue?

@namco1992 Sorry for the late response!
Could you please try the following procedures:

1. Re-clone your fork to the $GOPATH/src/github.com/pingcap/tidb-operator
  I know that this should be unnecessary as we have adopted the go module, however, it seems that some tools are still dependent on the GOPATH.
2. Set env for `GOBIN`, `GOROOT`, `GOPATH`

That what I am doing and I can generate the code as expected.

**Comment by namco1992** [comment]
> > @DanielZhangQD Unfortunately I faced the same issue as @sstubbs in #3134. I'm not able to regenerate the deepcopy and tried whatever @sstubbs tried too. Does any of you know how can I solve this issue?
> 
> @namco1992 Sorry for the late response!
> Could you please try the following procedures:
> 
> 1. Re-clone your fork to the $GOPATH/src/github.com/pingcap/tidb-operator
>    I know that this should be unnecessary as we have adopted the go module, however, it seems that some tools are still dependent on the GOPATH.
> 2. Set env for `GOBIN`, `GOROOT`, `GOPATH`
> 
> That what I am doing and I can generate the code as expected.

Thank you @DanielZhangQD move the repo under `$GOPATH` solves the problem!

**Comment by namco1992** [comment]
@DanielZhangQD I added a check for TiKV version and fail the backup/restore if the TiKV version < 4.0.8 and the `spec.from/to` is not set. However, then I realized that it's the BR version that need to be > 4.0.8, do I need to keep the check? or should I check the BR version instead?

**Comment by DanielZhangQD** [comment]
> @DanielZhangQD I added a check for TiKV version and fail the backup/restore if the TiKV version < 4.0.8 and the `spec.from/to` is not set. However, then I realized that it's the BR version that need to be > 4.0.8, do I need to keep the check? or should I check the BR version instead?

@namco1992 It's OK to keep the check, we will have PR to make sure that the version of BR matches with that of TiKV soon.

File: pkg/backup/util/util.go [hunk 1]
```diff
@@ -455,3 +474,17 @@ func ParseImage(image string) (string, string) {
...
 	return name, tag
 }
+
+// canSkipSetGCLifeTime returns if setting tikv_gc_life_time can be skipped based on the TiKV version
+func canSkipSetGCLifeTime(image string) bool {
+	_, version := ParseImage(image)
+	v, err := semver.NewVersion(version)
+	if err != nil {
+		klog.Errorf("Parse version %s failure, error: %v", version, err)
+		return false
```
Comment (DanielZhangQD) [inline pkg/backup/util/util.go:484]: `latest`, `nightly` are invalid Semantic Version, I think we can return true here.

File: pkg/backup/util/util.go [hunk 2]
```diff
@@ -385,23 +399,28 @@ func ValidateBackup(backup *v1alpha1.Backup) error {
...
+			} else {
+				if restore.Spec.To.Host == "" {
+					return fmt.Errorf("missing cluster config in spec of %s/%s", ns, name)
+				}
+
+				if restore.Spec.To.SecretName == "" {
+					return fmt.Errorf("missing tidbSecretName config in spec of %s/%s", ns, name)
+				}
+			}
+		}
```
Comment (DanielZhangQD) [inline pkg/backup/util/util.go:423]: We should move this block out of the `else` block as we have done for backup.

File: cmd/backup-manager/app/import/manager.go [hunk 3]
```diff
@@ -51,24 +47,6 @@ func NewRestoreManager(
 	}
 }
 
-func (rm *RestoreManager) setOptions(restore *v1alpha1.Restore) {
```
Comment (DanielZhangQD) [inline cmd/backup-manager/app/import/manager.go:54]: We also have to check the nil of `backup.spec.from` in the `cmd/backup-manager/app/export/manager.go`.

File: pkg/backup/restore/restore_manager.go [hunk 4]
```diff
@@ -135,9 +148,16 @@ func (rm *restoreManager) makeImportJob(restore *v1alpha1.Restore) (*batchv1.Job
...
 
-	envVars, reason, err := backuputil.GenerateTidbPasswordEnv(ns, name, restore.Spec.To.SecretName, restore.Spec.UseKMS, rm.deps.KubeClientset)
-	if err != nil {
-		return nil, reason, err
+	var (
+		envVars []corev1.EnvVar
+		reason  string
+		err     error
+	)
+	if restore.Spec.To != nil {
```
Comment (DanielZhangQD) [inline pkg/backup/restore/restore_manager.go:156]: We have removed the GC logic in `cmd/backup-manager/app/import/manager.go`, should we also remove this?

**Comment by DanielZhangQD** [comment]
> > @DanielZhangQD I added a check for TiKV version and fail the backup/restore if the TiKV version < 4.0.8 and the `spec.from/to` is not set. However, then I realized that it's the BR version that need to be > 4.0.8, do I need to keep the check? or should I check the BR version instead?
> 
> @namco1992 It's OK to keep the check, we will have PR to make sure that the version of BR matches with that of TiKV soon.

@namco1992 Since you have checked the version for BR, what about applying the following logic:

- Dumpling, if TiKV >= v4.0.0, no need to set `spec.from` (we can keep the similar logic with BR, if `spec.from` configured, go with the current GC logic, otherwise no need to set GC)
- BR, if TiKV >= v4.0.8, no need to set `spec.from` and `spec.to`
- Lightning, no need to set `spec.to`

cc @lichunzhu Please correct me if anything wrong.

File: pkg/backup/util/util.go [hunk 5]
```diff
@@ -326,16 +334,22 @@ func GetBackupDataPath(provider v1alpha1.StorageProvider) (string, string, error
...
+			if backup.Spec.From.Host == "" {
+				return fmt.Errorf("missing cluster config in spec of %s/%s", ns, name)
+			}
 
-	if backup.Spec.From.SecretName == "" {
-		return fmt.Errorf("missing tidbSecretName config in spec of %s/%s", ns, name)
+			if backup.Spec.From.SecretName == "" {
+				return fmt.Errorf("missing tidbSecretName config in spec of %s/%s", ns, name)
+			}
+		}
```
Comment (DanielZhangQD) [inline pkg/backup/util/util.go:352]: We can define this block as validateAccessConfig, and then:
if backup.Spec.BR == nil, we should call this function;
else if !canSkipSetGCLifeTime(tikvImage), call this.

**Comment by DanielZhangQD** [comment]
@namco1992 Did you test the backup and restore with `Dumpling/Lightning` and `BR` with your code change?

**Comment by namco1992** [comment]
@DanielZhangQD I tested with `BR` but not `Dumpling/Lightning`, I will address all your comments first and then test on `Dumpling/Lightning`. Thanks for your review!

**Comment by DanielZhangQD** [comment]
> @DanielZhangQD I tested with `BR` but not `Dumpling/Lightning`, I will address all your comments first and then test on `Dumpling/Lightning`. Thanks for your review!

OK. Thanks! I have made some wrong comments this morning and I have removed them, sorry for the confusion!

File: pkg/backup/backup/backup_manager.go [hunk 6]
```diff
@@ -65,7 +66,26 @@ func (bm *backupManager) syncBackupJob(backup *v1alpha1.Backup) error {
 	name := backup.GetName()
 	backupJobName := backup.GetBackupJobName()
 
-	err := backuputil.ValidateBackup(backup)
+	var err error
+	if backup.Spec.BR == nil {
```
Comment (namco1992) [inline pkg/backup/backup/backup_manager.go:70]: @DanielZhangQD you commented you check if `backup.Spec.From` is nil in the `cmd/backup-manager/app/export/manager.go`, however, we validated the `backup.Spec.From` here and make sure  it exists. Then do we still need the check?

File: pkg/backup/util/util.go [hunk 7]
```diff
@@ -326,23 +334,38 @@ func GetBackupDataPath(provider v1alpha1.StorageProvider) (string, string, error
...
+			return "missing tidbSecretName config in spec of %s/%s"
+		}
 	}
+	return ""
+}
 
-	if backup.Spec.From.SecretName == "" {
-		return fmt.Errorf("missing tidbSecretName config in spec of %s/%s", ns, name)
-	}
+func ValidateBackup(backup *v1alpha1.Backup, tikvImage string) error {
```
Comment (namco1992) [inline pkg/backup/util/util.go:352]: @DanielZhangQD I refactored the validation logic:
1. If the `backup.Spec.BR` is nil, the `TiDBAccessConfig` will be validated;
2. If the `backup.Spec.BR` presents and `!canSkipSetGCLifeTime(tikvImage)`, the `TiDBAccessConfig` will be validated.

**Comment by DanielZhangQD** [comment]
@namco1992 Are you available to check the comments today?
We can fix the issue https://github.com/pingcap/tidb-operator/issues/3441 based on your code change in this PR.

**Comment by namco1992** [comment]
> @namco1992 Are you available to check the comments today?
> We can fix the issue #3441 based on your code change in this PR.

Hello @DanielZhangQD I just pushed my changes and it passed the tests. Can help review again?

**Comment by DanielZhangQD** [comment]
> > @namco1992 Are you available to check the comments today?
> > We can fix the issue #3441 based on your code change in this PR.
> 
> Hello @DanielZhangQD I just pushed my changes and it passed the tests. Can help review again?

@namco1992 OK, sorry that I missed it, I will review the changes now.
BTW, are you available to take #3441?

**Comment by namco1992** [comment]
> BTW, are you available to take #3441?

Yes I would love to. But I don't have access to the proposal link, can you grant me access?

**Comment by DanielZhangQD** [comment]
> > BTW, are you available to take #3441?
> 
> Yes I would love to. But I don't have access to the proposal link, can you grant me access?

Please try again.
Thank you very much!
Add it would be much better if you can finish it by this Thursday and then we can release the fix on Friday.

**Comment by DanielZhangQD** [comment]
@namco1992 I also have some updates here https://github.com/pingcap/tidb-operator/issues/3441#issuecomment-720895950, please check if any questions.

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Review by lichunzhu — APPROVED** [review]
LGTM

**Comment by DanielZhangQD** [comment]
> @namco1992 Did you test the backup and restore with `Dumpling/Lightning` and `BR` with your code change?

@namco1992 You have done the tests mentioned above, right?

**Comment by namco1992** [comment]
@DanielZhangQD Yes I've tested on both.

**Pull request merged**